balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -456,30 +473,73 @@ func (b *pickfirstBalancer) requestConnectionLocked() {
 		switch scd.state {
 		case connectivity.Idle:
 			scd.subConn.Connect()
+			b.scheduleNextConnectionLocked()
+			return
 		case connectivity.TransientFailure:
-			// Try the next address.
+			// The SubConn is being re-used and failed during a previous pass
+			// over the addressList. It has not completed backoff yet.
+			// Mark it as having failed and try the next address.
+			scd.connectionFailed = true
 			lastErr = scd.lastErr
 			continue
 		case connectivity.Ready:
 			// Should never happen.
 			b.logger.Errorf("Requesting a connection even though we have a READY SubConn")
+			return
 		case connectivity.Shutdown:
 			// Should never happen.
 			b.logger.Errorf("SubConn with state SHUTDOWN present in SubConns map")
+			return
 		case connectivity.Connecting:
-			// Wait for the SubConn to report success or failure.
+			// Wait for the connection attempt to complete or the timer to fire
+			// before attempting the next address.
+			b.scheduleNextConnectionLocked()
```

Comment:
It feels like either this or the one in the `Idle` case is redundant.

Reply:
OK, I see.  I thought this was called from `updateSubConnState` -- and it is, but only in the case where the subchannel update was a TF update.  Thanks!